### ISSUE TITLE
Exception when navigating from a page with params to a page whose URL has less segments than the current one

### ISSUE THREAD

**Issue by wooky** [kwebio/kweb-core#102]
Consider the following example:
```kotlin
import io.kweb.Kweb
import io.kweb.dom.element.creation.tags.a
import io.kweb.dom.element.creation.tags.h1
import io.kweb.dom.element.new
import io.kweb.routing.route
import io.kweb.state.path
import io.kweb.state.simpleUrlParser

fun main() {
    Kweb(port = 13579) {
        doc.body.new {
            val uri = url(simpleUrlParser).path
            route {
                path("/this/has/four/{segments}") {
                    h1().text("This has four segments")
                    a(href = null).text("Go to three segments").on.click { uri.value = "/this/has/three" }
                }
                path("/this/has/three") {
                    h1().text("This has three segments")
                }
            }
        }
    }
}
```

When I go to `/this/has/four/whatever`, I get the text "This has four segment" and a link which goes to `/this/has/three`. Clicking on it does not navigate to that page, and instead I get this message in the console:
```
21:10:18.021 [qtp1552978964-31] ERROR io.kweb.dom.element.events.ONReceiver - Exception thrown by callback in response to click event
java.lang.IndexOutOfBoundsException: Index 3 out of bounds in list [this, has, three]
	at io.kweb.state.KVarExtensionsKt$get$1.invoke(KVarExtensions.kt:11) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVarExtensionsKt$get$1.invoke(KVarExtensions.kt:6) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:34) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:14) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$$special$$inlined$observable$1.afterChange(Delegates.kt:77) ~[kweb-core-0.5.16.jar:?]
	at kotlin.properties.ObservableProperty.setValue(ObservableProperty.kt:41) ~[kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at io.kweb.state.KVar.setValue(KVar.kt) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:34) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:14) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$$special$$inlined$observable$1.afterChange(Delegates.kt:77) ~[kweb-core-0.5.16.jar:?]
	at kotlin.properties.ObservableProperty.setValue(ObservableProperty.kt:41) ~[kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at io.kweb.state.KVar.setValue(KVar.kt) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:34) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:14) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$$special$$inlined$observable$1.afterChange(Delegates.kt:77) ~[kweb-core-0.5.16.jar:?]
	at kotlin.properties.ObservableProperty.setValue(ObservableProperty.kt:41) ~[kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at io.kweb.state.KVar.setValue(KVar.kt) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$map$origChangeHandle$1.invoke(KVar.kt:39) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$map$origChangeHandle$1.invoke(KVar.kt:14) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$$special$$inlined$observable$1.afterChange(Delegates.kt:77) ~[kweb-core-0.5.16.jar:?]
	at kotlin.properties.ObservableProperty.setValue(ObservableProperty.kt:41) ~[kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at io.kweb.state.KVar.setValue(KVar.kt) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$map$origChangeHandle$1.invoke(KVar.kt:39) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$map$origChangeHandle$1.invoke(KVar.kt:14) ~[kweb-core-0.5.16.jar:?]
	at io.kweb.state.KVar$$special$$inlined$observable$1.afterChange(Delegates.kt:77) ~[kweb-core-0.5.16.jar:?]
	at kotlin.properties.ObservableProperty.setValue(ObservableProperty.kt:41) ~[kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at io.kweb.state.KVar.setValue(KVar.kt) ~[kweb-core-0.5.16.jar:?]
	at UhKt$main$1$1$1$1$1.invoke(uh.kt:16) ~[classes/:?]
	at UhKt$main$1$1$1$1$1.invoke(uh.kt) ~[classes/:?]
	at io.kweb.dom.element.events.ONReceiver$click$$inlined$event$1.invoke(ONReceiver.kt:48) [kweb-core-0.5.16.jar:?]
	at io.kweb.dom.element.events.ONReceiver$click$$inlined$event$1.invoke(ONReceiver.kt:13) [kweb-core-0.5.16.jar:?]
	at io.kweb.dom.element.events.ONReceiver$event$1.invoke(ONReceiver.kt:38) [kweb-core-0.5.16.jar:?]
	at io.kweb.dom.element.events.ONReceiver$event$1.invoke(ONReceiver.kt:13) [kweb-core-0.5.16.jar:?]
	at io.kweb.dom.element.Element$addEventListener$1.invoke(Element.kt:286) [kweb-core-0.5.16.jar:?]
	at io.kweb.dom.element.Element$addEventListener$1.invoke(Element.kt:31) [kweb-core-0.5.16.jar:?]
	at io.kweb.Kweb$listenForWebsocketConnection$1.invokeSuspend(Kweb.kt:237) [kweb-core-0.5.16.jar:?]
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) [kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241) [kotlinx-coroutines-core-1.3.2-1.3.60-eap-76.jar:?]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806) [jetty-util-9.4.24.v20191120.jar:9.4.24.v20191120]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938) [jetty-util-9.4.24.v20191120.jar:9.4.24.v20191120]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
```

This is a very specific, but not unreasonable, case:
1. The current page has 4 segments, but one of them is a captured parameter. It doesn't matter if it gets queried or not.
2. The page we're going to has less segments than the current, and this causes issues. Doesn't matter how many less segments there are, or if any of them are captured parameters.
3. If we're going the opposite way, i.e. to a page with more segments than the current one, there are no issues.
4. If the current page has more segments than the next, but there are no captured parameters, there are no issues.

**Comment by sanity** [comment]
Hmm, I've only looked at this briefly but my initial guess is a problem with the logic in `routing.kt` where KVars created at [routing.kt:57](https://github.com/kwebio/kweb-core/blob/master/src/main/kotlin/io/kweb/routing/routing.kt#L57) aren't being cleaned up correctly when the `matchingTemplate` changes.  I will investigate further.

**Comment by sanity** [comment]
@wooky I believe this is fixed in 0.5.17, would appreciate it if you could verify and close this issue if it looks good.

**Comment by wooky** [comment]
It's navigating fine now, but the exception message is still being logged.

**Comment by sanity** [comment]
Gotcha, I'll need to do more debugging.

**Comment by sanity** [comment]
@wooky Would you mind sending me the exception you're getting with 0.5.17?  Was having some trouble reproducing.

**Comment by wooky** [comment]
It appears to be the same exception, stack trace and all, but is now logged as a warning rather than an error:
```
18:26:01.103 [qtp199640888-39] WARN  io.kweb.state.KVar - Exception thrown by listener
java.lang.IndexOutOfBoundsException: Index 3 out of bounds in list [this, has, three]
	at io.kweb.state.KVarExtensionsKt$get$1.invoke(KVarExtensions.kt:11) ~[kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVarExtensionsKt$get$1.invoke(KVarExtensions.kt:6) ~[kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:37) ~[kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:14) ~[kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$$special$$inlined$observable$1.afterChange(Delegates.kt:78) [kweb-core-0.5.17.jar:?]
	at kotlin.properties.ObservableProperty.setValue(ObservableProperty.kt:41) [kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at io.kweb.state.KVar.setValue(KVar.kt) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:37) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:14) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$$special$$inlined$observable$1.afterChange(Delegates.kt:78) [kweb-core-0.5.17.jar:?]
	at kotlin.properties.ObservableProperty.setValue(ObservableProperty.kt:41) [kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at io.kweb.state.KVar.setValue(KVar.kt) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:37) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$map$myChangeHandle$1.invoke(KVar.kt:14) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$$special$$inlined$observable$1.afterChange(Delegates.kt:78) [kweb-core-0.5.17.jar:?]
	at kotlin.properties.ObservableProperty.setValue(ObservableProperty.kt:41) [kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at io.kweb.state.KVar.setValue(KVar.kt) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$map$origChangeHandle$1.invoke(KVar.kt:43) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$map$origChangeHandle$1.invoke(KVar.kt:14) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$$special$$inlined$observable$1.afterChange(Delegates.kt:78) [kweb-core-0.5.17.jar:?]
	at kotlin.properties.ObservableProperty.setValue(ObservableProperty.kt:41) [kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at io.kweb.state.KVar.setValue(KVar.kt) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$map$origChangeHandle$1.invoke(KVar.kt:43) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$map$origChangeHandle$1.invoke(KVar.kt:14) [kweb-core-0.5.17.jar:?]
	at io.kweb.state.KVar$$special$$inlined$observable$1.afterChange(Delegates.kt:78) [kweb-core-0.5.17.jar:?]
	at kotlin.properties.ObservableProperty.setValue(ObservableProperty.kt:41) [kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at io.kweb.state.KVar.setValue(KVar.kt) [kweb-core-0.5.17.jar:?]
	at UhKt$main$1$1$1$1$1.invoke(uh.kt:19) [classes/:?]
	at UhKt$main$1$1$1$1$1.invoke(uh.kt) [classes/:?]
	at io.kweb.dom.element.events.ONReceiver$click$$inlined$event$1.invoke(ONReceiver.kt:48) [kweb-core-0.5.17.jar:?]
	at io.kweb.dom.element.events.ONReceiver$click$$inlined$event$1.invoke(ONReceiver.kt:13) [kweb-core-0.5.17.jar:?]
	at io.kweb.dom.element.events.ONReceiver$event$1.invoke(ONReceiver.kt:38) [kweb-core-0.5.17.jar:?]
	at io.kweb.dom.element.events.ONReceiver$event$1.invoke(ONReceiver.kt:13) [kweb-core-0.5.17.jar:?]
	at io.kweb.dom.element.Element$addEventListener$1.invoke(Element.kt:286) [kweb-core-0.5.17.jar:?]
	at io.kweb.dom.element.Element$addEventListener$1.invoke(Element.kt:31) [kweb-core-0.5.17.jar:?]
	at io.kweb.Kweb$listenForWebsocketConnection$1.invokeSuspend(Kweb.kt:237) [kweb-core-0.5.17.jar:?]
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) [kotlin-stdlib-1.3.60.jar:1.3.60-release-155 (1.3.60)]
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241) [kotlinx-coroutines-core-1.3.2-1.3.60-eap-76.jar:?]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806) [jetty-util-9.4.24.v20191120.jar:9.4.24.v20191120]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938) [jetty-util-9.4.24.v20191120.jar:9.4.24.v20191120]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
```

**Comment by sanity** [comment]
@wooky Thank you, I think the solution here might be to make `KVar.map()` close the mapped `KVar` in the event of the mapper throwing an exception, along with a mechanism to record why KVars were closed.

**Comment by sanity** [comment]
@wooky If you get a chance, would you mind trying `implementation 'com.github.kwebio:kweb-core:-SNAPSHOT' to see whether my recent commit fixes that exception?  If so I'll do another release.

**Comment by wooky** [comment]
I'm not able to use it. Maven is complaining about not being able to find:
- org.jetbrains.kotlinx:kotlinx-coroutines-core:jar:1.3.3-1.3.61-eap-76
- org.jetbrains.kotlinx:kotlinx-coroutines-jdk8:jar:1.3.3-1.3.61-eap-76

**Comment by sanity** [comment]
Hmm, that's odd, ok - that dependency issue should be fixed in version `0.5.19` which I've just released, would appreciate it if you could try it out and let me know how it goes.

**Comment by wooky** [comment]
I no longer get any errors. Thanks!
